source/slang/slang-ir-lower-defer.cpp

```diff
@@ -0,0 +1,256 @@
+        // instructions at the target position, in the existing block.
+        if (deferBlocks.getCount() == 1)
+        {
+            inlineSingleBlockDefer(beforeInst, deferBlocks.getFirst(), builder);
+            return targetBlock;
+        }
+
+        // Otherwise, we'll have to splice the blocks in.
+        IRCloneEnv env;
+        builder->setInsertAfter(targetBlock);
+        auto lastBlock = targetBlock;
+
+        // Clone blocks first
+        for (auto block : deferBlocks)
+        {
+            auto clonedBlock = builder->createBlock();
+            builder->addInst(clonedBlock);
+            env.mapOldValToNew[block] = clonedBlock;
+        }
+
+        // Then, clone instructions, but mapping old blocks to new blocks.
+        for (auto block : deferBlocks)
+        {
+            auto clonedBlock = as<IRBlock>(env.mapOldValToNew.getValue(block));
+            builder->setInsertInto(clonedBlock);
+            for (auto inst : block->getChildren())
+            {
+                auto endBranch = as<IRUnconditionalBranch>(inst);
+                if (endBranch && endBranch->getTargetBlock() == mergeBlock)
+                {
+                    lastBlock = clonedBlock;
+                }
+                else
+                    cloneInst(&env, builder, inst);
+            }
+        }
+
+        // Move old instructions to the last block's end. The last defer block
+        // shouldn't have a terminator at this point yet.
+        while (beforeInst)
+        {
+            auto nextInst = beforeInst->getNextInst();
+            beforeInst->insertAtEnd(lastBlock);
+            beforeInst = nextInst;
+        }
+
+        // Make target block jump to the cloned blocks.
+        builder->setInsertInto(targetBlock);
+        auto mainBlock = as<IRBlock>(env.mapOldValToNew.getValue(deferBlocks[0]));
+        builder->emitBranch(mainBlock);
+
+        return lastBlock;
+    }
+
+    bool isBlockInScope(IRBlock* block, IRDominatorTree* dom, IRBlock* scopeEndBlock)
+    {
+        if (block == scopeEndBlock)
+            return false;
+
+        // Technically, we'd like to know if 'scopeEndBlock' is a predecessor of
+        // 'block'. This is usually the same as 'scopeEndBlock' dominating the
+        // given block, except for when a statement can jump over the scope end
+        // without completely exiting the dominance of the 'defer' statement
+        // itself. The only case like this is the continue statement of a for
+        // loop.
+        if (dom->properlyDominates(scopeEndBlock, block))
+            return false;
+
+        // The continuation block is not always dominated by the end of the
+        // scope, since a continue statement is another way to get there.
+        // Luckily, we can just check if the scope end is directly succeeded by
+        // the target block, which is the case in for loops.
+        for (auto successor : scopeEndBlock->getSuccessors())
```

Comment:
I wonder if we can get rid of this work around, if we handle the situation in lower-to-ir.

If you are lowering a block statement directly inside a loop, then don’t create a dedicated scopeEnd block, just use the continueBlock of the loop.